src/Abstractions/NexusMods.Abstractions.GameLocators/LocationId.cs

```diff
@@ -1,11 +1,48 @@
-﻿namespace NexusMods.Abstractions.GameLocators;
+﻿using System.Collections.Immutable;
+using TransparentValueObjects;
+
+namespace NexusMods.Abstractions.GameLocators;
 
 /// <summary>
 /// The base folder for the GamePath, more values can easily be added here as needed
 /// </summary>
-[TransparentValueObjects.ValueObject<string>]
+[ValueObject<ushort>]
 public readonly partial struct LocationId
 {
+    private static ImmutableDictionary<ushort, string> _cache = ImmutableDictionary<ushort, string>.Empty;
+    
+    /// <summary>
+    /// Converts the string to a LocationId, if the string is not already interned it will be added to the cache, if a hash collision is
+    /// detected (highly unlikely) an exception will be thrown.
+    /// </summary>
+    /// <param name="value"></param>
+    /// <returns></returns>
+    /// <exception cref="InvalidOperationException"></exception>
+    public static LocationId From(string value)
+    {
+        var hash = FNV1aHash.Hash(value);
+        var mixed = (ushort) ((hash >> 16) ^ (hash & 0xFFFF));
+        if (_cache.TryGetValue(mixed, out var existing))
+        {
+            if (existing != value)
+            {
+                throw new InvalidOperationException($"Hash collision detected for '{value}' and '{existing}'");
+            }
+            return From(mixed);
+        }
+        
+        while (true)
+        {
+            var newCache = _cache.Add(mixed, value);
+            if (ReferenceEquals(Interlocked.CompareExchange(ref _cache, newCache, _cache), _cache))
+                break;
+        }
```

Comment:
I'm curious, what is the user supposed to do if they encounter this? Wait until we implement hash collision resolution? If this is highly unlikely to happen anyways, we can move the check into debug mode.

Reply:
My thought is that whomever is implementing a new game, will hit this, then slightly alter the name of the location id, just enough to get it not to collide. So we're likely only going to get collisions during the initial steps of adding a new game.